src/core/meshcop/border_agent.cpp

```diff
@@ -308,23 +309,40 @@ Error BorderAgent::SetId(const Id &aId)
 
 void BorderAgent::HandleNotifierEvents(Events aEvents)
 {
-    VerifyOrExit(aEvents.ContainsAny(kEventThreadRoleChanged | kEventCommissionerStateChanged));
-
+    if (aEvents.ContainsAny(kEventThreadRoleChanged | kEventCommissionerStateChanged))
+    {
 #if OPENTHREAD_CONFIG_COMMISSIONER_ENABLE && OPENTHREAD_FTD
-    VerifyOrExit(Get<Commissioner>().IsDisabled());
+        VerifyOrExit(Get<Commissioner>().IsDisabled());
 #endif
 
-    if (Get<Mle::MleRouter>().IsAttached())
-    {
-        Start();
+        if (Get<Mle::MleRouter>().IsAttached())
+        {
+            Start();
+        }
+        else
+        {
+            Stop();
+        }
     }
-    else
+
+    if (aEvents.ContainsAny(kEventPskcChanged))
     {
-        Stop();
+        if (mState == kStateStarted)
+        {
+            Restart(kRestartOnPskcChanged);
+        }
+
+        mPskcChanged = true;
+
+        // For active state (kStateConnected or kStateAccepted), there
+        // will be restart after secure session is disconnected, thus
+        // giving it the chance to apply new pskc.
     }
 
+#if OPENTHREAD_CONFIG_COMMISSIONER_ENABLE && OPENTHREAD_FTD
```

Comment:
Can consider adding an `ExitNow()` at the end, then we don't need `#if` around the `exit` label?